platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/ExcludingVariantArtifactSet.java

```diff
@@ -0,0 +1,114 @@
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariant;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ExcludeSpec;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.model.VariantResolveMetadata;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A {@link ResolvedVariant} that applies artifact exclusions to a delegate {@link ResolvedVariant}.
+ */
+public class ExcludingVariantArtifactSet implements ResolvedVariant, VariantResolveMetadata.Identifier {
+
+    private final ResolvedVariant delegate;
+    private final ModuleIdentifier moduleId;
+    private final ExcludeSpec exclusions;
+
+    public ExcludingVariantArtifactSet(ResolvedVariant delegate, ModuleIdentifier moduleId, ExcludeSpec exclusions) {
+        this.delegate = delegate;
+        this.moduleId = moduleId;
+        this.exclusions = exclusions;
+    }
+
+    @Override
+    public DisplayName asDescribable() {
+        return Describables.of(delegate.asDescribable(), exclusions);
+    }
+
+    @Nullable
+    @Override
+    public VariantResolveMetadata.Identifier getIdentifier() {
+        // Since we use ourselves as the identifier, we must implement equals and hashCode
+        return this;
+    }
+
+    @Override
+    public ImmutableAttributes getAttributes() {
+        return delegate.getAttributes();
+    }
+
+    @Override
+    public ImmutableCapabilities getCapabilities() {
+        return delegate.getCapabilities();
+    }
+
+    @Override
+    public ResolvedArtifactSet getArtifacts() {
+        ResolvedArtifactSet artifacts = delegate.getArtifacts();
+        return new FilteringResolvedArtifactSet(artifacts, this::include);
+    }
+
+    private boolean include(ResolvableArtifact artifact) {
+        return !exclusions.excludesArtifact(moduleId, artifact.getArtifactName());
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hashCode(delegate.getIdentifier());
+        result = 31 * result + moduleId.hashCode();
+        result = 31 * result + exclusions.hashCode();
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        ExcludingVariantArtifactSet other = (ExcludingVariantArtifactSet) obj;
+        return equalId(delegate.getIdentifier(), other.delegate.getIdentifier()) &&
+            moduleId.equals(other.moduleId) &&
+            exclusions.equals(other.exclusions);
+    }
+
+    private static boolean equalId(
```

Comment:
🤔 Shouldn't this live in `VariantResolveMetadata.Identifier` itself, as `equals()`?  Then we could just use `Objects.equals()` in this type's `equals()` for the identifier comparison?

Reply:
It cannot, this method operates on nullable instances